kalviumcommunity/S48_Class-Lazy-Workers
# Joi validation and created login page

# Pull Request Description

## What does this PR do?

This PR introduces a Joi validation schema for validating user data in the `user.js` file. The schema defines the expected shape and data types for user objects, ensuring that only valid data is stored in the MongoDB database.

## Why is this change necessary?

Currently, the `UserSchema` defined in `user.js` does not have any validation rules, allowing potentially invalid or malformed data to be stored in the database. By implementing a robust validation schema using Joi, we can enforce data integrity and consistency, improving the overall reliability and maintainability of our application.

## How was this implemented?

The new `userValidationSchema` is defined using Joi's schema builder functions. The schema includes validation rules for the following properties:

- `username` (required string, alphanumeric, minimum length of 3, maximum length of 30)
- `name` (required string, minimum length of 3, maximum length of 50)
- `email` (required string, must be a valid email address)
- `password` (required string, must match a specific regular expression pattern)
- `github` (optional string, must be a valid URI)
- `squad` (required number, integer, minimum value of 1, maximum value of 100)

The `userValidationSchema` is then exported alongside the `UserModel` for use in other parts of the application.

## How to test?

1. Start the server in development mode: `npm run dev`
2. Use a tool like Postman or cURL to send requests to create or update user data with various combinations of valid and invalid data.
3. Verify that valid requests are processed as expected, and invalid requests return appropriate error responses with detailed validation error messages.

## Additional Notes

- The schema can be easily extended or modified to accommodate future changes in the user data structure.
- Error messages are customized to provide more meaningful feedback to API consumers or internal processes.
- Input validation is a crucial step in ensuring data integrity and security, and this change helps improve our application's overall robustness.

## Issues Resolved

- Closes #37 (Implement user data validation using Joi)
